### PR TITLE
Libraries

### PR DESCRIPTION
**Pull request by zhangshyue**
Add authentication at comment page. User would be able to login, logout and set username. Login is now required for making comments. 
Screenshot 2020-07-06 at 8 41 35 PM

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/servlets/AccountServlet.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+        Gson gson = new Gson();
+        String json = new String();
+
+        // Get login information of the user
+        if (!userService.isUserLoggedIn()) {
+            String loginUrl = userService.createLoginURL("/comment.html");
+            status.add("Logout");
+            status.add(loginUrl);
+            json = gson.toJson(status);
+        }else{
```
Comment (wujingya) [inline portfolio/src/main/java/com/google/sps/servlets/AccountServlet.java]: nit: space around else

File: portfolio/src/main/java/com/google/sps/servlets/AccountServlet.java [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+            json = gson.toJson(status);
+        }
+        response.setContentType("application/json;");
+        response.getWriter().println(json);
+    }
+
+    /**
+    * Returns the username of the user with id, or empty String if the user has not set a username.
+    */
+    private String getUserUsername(String id) {
```
Comment (wujingya) [inline portfolio/src/main/java/com/google/sps/servlets/AccountServlet.java]: maybe just getUsername?

**Review by wujingya — COMMENTED** [review]
As we discussed, let's move the html stuff to the client side.

**Review by wujingya — APPROVED** [review]

File: portfolio/src/main/java/com/google/sps/servlets/AccountServlet.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+public class AccountServlet extends HttpServlet {
+
+    private List<String> status;
+
+    @Override
+    public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        response.setContentType("text/html");
+        PrintWriter out = response.getWriter();
+        UserService userService = UserServiceFactory.getUserService();
+        status = new ArrayList<>();
```
Comment (wujingya) [inline portfolio/src/main/java/com/google/sps/servlets/AccountServlet.java:42]: Can we create a class for status?

File: portfolio/src/main/webapp/comment.html [hunk 4]
```diff
@@ -11,7 +11,7 @@
         <script src="https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.14.7/umd/popper.min.js" integrity="sha384-UO2eT0CpHqdSJQ6hJty5KVphtPhzWj9WO1clHTMGa3JDZwrnQq4sF86dIHNDz0W1" crossorigin="anonymous"></script>
         <script src="https://stackpath.bootstrapcdn.com/bootstrap/4.3.1/js/bootstrap.min.js" integrity="sha384-JjSmVgyd0p3pXB1rRibZUAYoIIy6OrQ6VrjIEaFf/nJGzIxFDsf4x0xIM+B07jRM" crossorigin="anonymous"></script>
     </head>
-    <body onload="getContentFunctions()">
+    <body onload="getContentFunctions(); checkLogin();">
```
Comment (wujingya) [inline portfolio/src/main/webapp/comment.html]: How about make a onload() function that calls both of these, and then just call onload() here, it helps make things more readable, especially if you're adding more to this line later on.

Comment (zhangshyue) [reply portfolio/src/main/webapp/comment.html]: Sure!

**Pull request merged**